configmerge/configmerge.go

```diff
@@ -0,0 +1,244 @@
+		m.repoInfo = repoInfo
+	}
+
+	mainConfigRef := ConfigReference{
+		Path: mainConfigPth,
+	}
+
+	if repoInfo != nil {
+		mainConfigRef.Repository = repoInfo.DefaultRemoteURL
+		mainConfigRef.Commit = repoInfo.Commit
+		mainConfigRef.Tag = repoInfo.Tag
+		mainConfigRef.Branch = repoInfo.Branch
+	}
+
+	mainConfigBytes, err := m.fileReader.ReadFileFromFileSystem(mainConfigPth)
+	if err != nil {
+		return "", nil, err
+	}
+
+	mainConfigDir := filepath.Dir(mainConfigPth)
+	configTree, err := m.buildConfigTree(mainConfigBytes, mainConfigRef, mainConfigDir, 1, nil)
+	if err != nil {
+		return "", nil, err
+	}
+
+	mergedConfigContent, err := configTree.Merge()
+	if err != nil {
+		return "", nil, err
+	}
+
+	return mergedConfigContent, configTree, nil
+}
+
+func (m *Merger) buildConfigTree(configContent []byte, reference ConfigReference, dir string, depth int, keys []string) (*models.ConfigFileTreeModel, error) {
+	key := reference.Key()
+	keys = append(keys, key)
+
+	if len(configContent) > MaxFileSizeBytes {
+		return nil, fmt.Errorf("max file size (%d bytes) exceeded in file %s", MaxFileSizeBytes, key)
+	}
+
+	if depth > MaxIncludeDepth {
+		return nil, fmt.Errorf("max include depth (%d) exceeded", MaxIncludeDepth)
+	}
+
+	m.filesCount++
+	if m.filesCount > MaxFilesCountTotal {
+		return nil, fmt.Errorf("max include count (%d) exceeded", MaxFilesCountTotal)
+	}
+
+	var config struct {
+		Include []ConfigReference `yaml:"include" json:"include"`
+	}
+	if err := yaml.Unmarshal(configContent, &config); err != nil {
+		return nil, err
+	}
+
+	if len(config.Include) > MaxIncludeCountPerFile {
+		return nil, fmt.Errorf("max include count (%d) exceeded", MaxIncludeCountPerFile)
+	}
+	if m.filesCount+len(config.Include) > MaxFilesCountTotal {
+		return nil, fmt.Errorf("max file count (%d) exceeded", MaxFilesCountTotal)
+	}
+
+	for idx, include := range config.Include {
+		if err := include.Validate(); err != nil {
+			return nil, err
+		}
+		if include.Repository == "" {
+			include.Repository = reference.Repository
+			include.Branch = reference.Branch
+			include.Commit = reference.Commit
+			include.Tag = reference.Tag
+		}
+		config.Include[idx] = include
+
+		if sliceutil.IsStringInSlice(include.Key(), keys) {
+			return nil, fmt.Errorf("circular reference detected: %s -> %s", strings.Join(keys, " -> "), include.Key())
+		}
+	}
+
+	var includedConfigTrees []models.ConfigFileTreeModel
+	for _, include := range config.Include {
+		moduleBytes, err := m.readConfigModule(include, dir, m.repoInfo)
+		if err != nil {
+			return nil, err
+		}
+
+		moduleDir := filepath.Dir(include.Path)
+		moduleConfigTree, err := m.buildConfigTree(moduleBytes, include, moduleDir, depth+1, keys)
+		if err != nil {
+			return nil, err
+		}
+
+		includedConfigTrees = append(includedConfigTrees, *moduleConfigTree)
+	}
+
+	return &models.ConfigFileTreeModel{
+		Path:     key,
+		Contents: string(configContent),
+		Includes: includedConfigTrees,
+		Depth:    depth,
+	}, nil
```

Comment:
Would it help if we flatten this tree before we actually merge it? If we flatten it to an array then the merge is simply iterating over the items one by one. And at the moment it is recursive `ConfigFileTreeModel.Merge` function.

Reply:
I'm not sure if this would make the code simpler, because we anyway need to traverse the tree first.